File: ios/HybridBleScanner.swift
//
//  HybridBleScanner.swift
//  react-native-ble-print-and-scan
//
//  Created by Claude Code on 2025-01-05.
//

import Foundation
import UIKit
import NitroModules

class HybridBleScanner: HybridBleScannerSpec {
    
    private var scannerManager: ScannerBluetoothManager? = nil
    
    // MARK: - Scanner Connection Management
    
    func initializeScanner() throws -> Promise<Void> {
        return Promise.async {
            if self.scannerManager == nil {
                self.scannerManager = ScannerBluetoothManager.shared
            }
            
            let bluetoothInitialized = self.scannerManager?.requestBluetoothPermissions() ?? false
            
            if !bluetoothInitialized {
                throw NSError(domain: "ScannerBluetoothError", code: 1, userInfo: [NSLocalizedDescriptionKey: "Bluetooth permissions not granted"])
            }
        }
    }
    
    func startScanningForScanners(onScannerFound: @escaping ([Device]) -> Void) throws -> Promise<Void> {
        return Promise.async {
            guard let manager = self.scannerManager else {
                throw NSError(domain: "ScannerBluetoothError", code: 2, userInfo: [NSLocalizedDescriptionKey: "ScannerBluetoothManager not initialized"])
            }
            
            guard manager.isBluetoothSupported() else {
                throw NSError(domain: "ScannerBluetoothError", code: 3, userInfo: [NSLocalizedDescriptionKey: "Bluetooth is not supported on this device"])
            }
            
            guard manager.isBluetoothEnabled() else {
                throw NSError(domain: "ScannerBluetoothError", code: 4, userInfo: [NSLocalizedDescriptionKey: "Please enable bluetooth"])
            }
            
            manager.startScanning(onScannerFound: onScannerFound)
        }
    }
    
    func suspendScanForScanners() throws -> Promise<Void> {
        return Promise.async {
            guard let manager = self.scannerManager else {
                throw NSError(domain: "ScannerBluetoothError", code: 2, userInfo: [NSLocalizedDescriptionKey: "ScannerBluetoothManager not initialized"])
            }
            
            _ = await manager.stopScanning()
        }
    }
    
    func connectToScanner(deviceId: String) throws -> Promise<Void> {
        return Promise.async {
            guard let manager = self.scannerManager else {
                throw NSError(domain: "ScannerBluetoothError", code: 2, userInfo: [NSLocalizedDescriptionKey: "ScannerBluetoothManager not initialized"])
            }
            
            guard manager.isBluetoothSupported() else {
                throw NSError(domain: "ScannerBluetoothError", code: 3, userInfo: [NSLocalizedDescriptionKey: "Bluetooth is not supported on this device"])
            }
            
            guard let uuid = UUID(uuidString: deviceId) else {
                throw NSError(domain: "ScannerBluetoothError", code: 5, userInfo: [NSLocalizedDescriptionKey: "Invalid device ID"])
            }
            
            _ = try await manager.connectToScanner(identifier: uuid)
        }
    }
    
    func disconnectFromScanner(deviceId: String) throws -> Promise<Void> {
        return Promise.async {
            guard let manager = self.scannerManager else {
                throw NSError(domain: "ScannerBluetoothError", code: 2, userInfo: [NSLocalizedDescriptionKey: "ScannerBluetoothManager not initialized"])
            }
            
            _ = try await manager.disconnect(deviceId: deviceId)
        }
    }
    
    // MARK: - Multi-Scanner Management
    
    func isScannerConnected(deviceId: String) throws -> Promise<Bool> {
        return Promise.async {
            guard let manager = self.scannerManager else {
                throw NSError(domain: "ScannerBluetoothError", code: 2, userInfo: [NSLocalizedDescriptionKey: "ScannerBluetoothManager not initialized"])
            }
            
            return manager.isConnected(deviceId: deviceId)
        }
    }
    
    func getConnectedScanners() throws -> Promise<[Device]> {
        return Promise.async {
            guard let manager = self.scannerManager else {
                throw NSError(domain: "ScannerBluetoothError", code: 2, userInfo: [NSLocalizedDescriptionKey: "ScannerBluetoothManager not initialized"])
            }
            
            return manager.getConnectedScanners()
        }
    }
    
    func disconnectAllScanners() throws -> Promise<Void> {
        return Promise.async {
            guard let manager = self.scannerManager else {
                throw NSError(domain: "ScannerBluetoothError", code: 2, userInfo: [NSLocalizedDescriptionKey: "ScannerBluetoothManager not initialized"])
            }
            
            _ = try await manager.disconnectAllScanners()
        }
    }
    
    // MARK: - Scanner Information
    
    func getScannerInfo(deviceId: String) throws -> Promise<ScannerInfo> {
        return Promise.async {
            guard let manager = self.scannerManager else {
                throw NSError(domain: "ScannerBluetoothError", code: 2, userInfo: [NSLocalizedDescriptionKey: "ScannerBluetoothManager not initialized"])
            }
            
            // Get firmware version
            let firmwareVersion = try await manager.sendCommand(deviceId: deviceId, command: "$SW#VER")
            
            return ScannerInfo(
                firmwareVersion: firmwareVersion,
                moduleType: "Unknown", // Could be retrieved with additional commands
                batteryLevel: nil, // Could be retrieved with "%BAT_VOL#"
                isConnected: manager.isConnected(deviceId: deviceId)
            )
        }
    }
    
    func getScannerSettings(deviceId: String) throws -> Promise<ScannerCurrentSettings> {
        return Promise.async {
            guard let manager = self.scannerManager else {
                throw NSError(domain: "ScannerBluetoothError", code: 2, userInfo: [NSLocalizedDescriptionKey: "ScannerBluetoothManager not initialized"])
            }
            
            // This would require multiple commands to get all settings
            // For now, return default settings
            return ScannerCurrentSettings(
                workMode: ScannerMode.hostTrigger,
                beepSettings: BeepSettings(
                    volume: BeepVolume.middle,
                    tone: BeepTone.highTone,
                    enabled: true,
                    customLevel: nil
                ),
                powerSettings: PowerSettings(
                    sleepTimeMinutes: 0,
                    autoSleepEnabled: false
                ),
                dataFormatSettings: DataFormatSettings(
                    prefixEnabled: false,
                    suffixEnabled: false,
                    hideBarcodePrefix: false,
                    hideBarcodeContent: false,
                    hideBarcodeSuffix: false
                ),
                timestampEnabled: false
            )
        }
    }
    
    // MARK: - Scan Operations
    
    func startListening(deviceId: String, onScanResult: @escaping (ScanResult) -> Void) throws -> Promise<Void> {
        return Promise.async {
            guard let manager = self.scannerManager else {
                throw NSError(domain: "ScannerBluetoothError", code: 2, userInfo: [NSLocalizedDescriptionKey: "ScannerBluetoothManager not initialized"])
            }
            
            try manager.startListening(deviceId: deviceId, onScanResult: onScanResult)
        }
    }
    
    func stopListening(deviceId: String) throws -> Promise<Void> {
        return Promise.async {
            guard let manager = self.scannerManager else {
                throw NSError(domain: "ScannerBluetoothError", code: 2, userInfo: [NSLocalizedDescriptionKey: "ScannerBluetoothManager not initialized"])
            }
            
            try manager.stopListening(deviceId: deviceId)
        }
    }
    
    func triggerScan(deviceId: String, duration: ScannerTrigger) throws -> Promise<Void> {
        return Promise.async {
            guard let manager = self.scannerManager else {
                throw NSError(domain: "ScannerBluetoothError", code: 2, userInfo: [NSLocalizedDescriptionKey: "ScannerBluetoothManager not initialized"])
            }
            
            let command = "%SCANTM#\(duration.rawValue)#"
            _ = try await manager.sendCommand(deviceId: deviceId, command: command)
        }
    }
    
    // MARK: - Scanner Configuration
    
    func setScannerMode(deviceId: String, mode: ScannerMode) throws -> Promise<Void> {
        return Promise.async {
            guard let manager = self.scannerManager else {
                throw NSError(domain: "ScannerBluetoothError", code: 2, userInfo: [NSLocalizedDescriptionKey: "ScannerBluetoothManager not initialized"])
            }
            
            let command: String
            switch mode {
            case .keyHold:
                command = "%SCMD#00#"
            case .continuous:
                command = "%SCMD#01#"
            case .keyPulse:
                command = "%SCMD#02#"
            case .hostTrigger:
                command = "%SCMD#03#"
            }
            
            _ = try await manager.sendCommand(deviceId: deviceId, command: command)
        }
    }
    
    func setBeepSettings(deviceId: String, settings: BeepSettings) throws -> Promise<Void> {
        return Promise.async {
            guard let manager = self.scannerManager else {
                throw NSError(domain: "ScannerBluetoothError", code: 2, userInfo: [NSLocalizedDescriptionKey: "ScannerBluetoothManager not initialized"])
            }
            
            let command: String
            switch settings.volume {
            case .mute:
                command = "$BUZZ#0"
            case .low:
                command = "$BUZZ#3"
            case .middle:
                command = "$BUZZ#2"
            case .high:
                command = "$BUZZ#1"
            }
            
            _ = try await manager.sendCommand(deviceId: deviceId, command: command)
        }
    }
    
    func setPowerSettings(deviceId: String, settings: PowerSettings) throws -> Promise<Void> {
        return Promise.async {
            guard let manager = self.scannerManager else {
                throw NSError(domain: "ScannerBluetoothError", code: 2, userInfo: [NSLocalizedDescriptionKey: "ScannerBluetoothManager not initialized"])
            }
            
            let command: String
            switch settings.sleepTimeMinutes {
            case 0:
                command = "$RF#ST00" // Never sleep
            case 1:
                command = "$RF#ST02"
            case 3:
                command = "$RF#ST06"
            case 5:
                command = "$RF#ST10"
            case 10:
                command = "$RF#ST20"
            case 30:
                command = "$RF#ST60"
            case 60:
                command = "$RF#ST<0"
            case 120:
                command = "$RF#STH0"
            default:
                command = "$RF#ST00" // Default to never sleep
            }
            
            _ = try await manager.sendCommand(deviceId: deviceId, command: command)
        }
    }
    
    func setDataFormatSettings(deviceId: String, settings: DataFormatSettings) throws -> Promise<Void> {
        return Promise.async {
            guard let manager = self.scannerManager else {
                throw NSError(domain: "ScannerBluetoothError", code: 2, userInfo: [NSLocalizedDescriptionKey: "ScannerBluetoothManager not initialized"])
            }
            
            var commands: [String] = []
            
            if settings.prefixEnabled {
                commands.append("$DATA#2")
            }
            if settings.suffixEnabled {
                commands.append("$DATA#1")
            }
            if settings.hideBarcodePrefix {
                commands.append("$DATA#5")
            }
            if settings.hideBarcodeContent {
                commands.append("$DATA#4")
            }
            if settings.hideBarcodeSuffix {
                commands.append("$DATA#3")
            }
            
            for command in commands {
                _ = try await manager.sendCommand(deviceId: deviceId, command: command)
            }
        }
    }
    
    func setTimestamp(deviceId: String, format: TimestampFormat, datetime: String?) throws -> Promise<Void> {
        return Promise.async {
            guard let manager = self.scannerManager else {
                throw NSError(domain: "ScannerBluetoothError", code: 2, userInfo: [NSLocalizedDescriptionKey: "ScannerBluetoothManager not initialized"])
            }
            
            let command: String
            switch format {
            case .disabled:
                command = "%RTCSTAMP#0"
            case .dateTime, .unixTimestamp:
                command = "%RTCSTAMP#1"
            }
            
            _ = try await manager.sendCommand(deviceId: deviceId, command: command)
        }
    }
    
    // MARK: - Advanced Commands
    
    func restoreFactorySettings(deviceId: String) throws -> Promise<Void> {
        return Promise.async {
            guard let manager = self.scannerManager else {
                throw NSError(domain: "ScannerBluetoothError", code: 2, userInfo: [NSLocalizedDescriptionKey: "ScannerBluetoothManager not initialized"])
            }
            
            _ = try await manager.sendCommand(deviceId: deviceId, command: "%#IFSNO$B")
        }
    }
    
    func customBeep(deviceId: String, level: Double) throws -> Promise<Void> {
        return Promise.async {
            guard let manager = self.scannerManager else {
                throw NSError(domain: "ScannerBluetoothError", code: 2, userInfo: [NSLocalizedDescriptionKey: "ScannerBluetoothManager not initialized"])
            }
            
            let clampedLevel = max(0, min(26, Int(level)))
            let command = "$BUZZ#B\(String(format: "%c", clampedLevel + 0x30))"
            
            _ = try await manager.sendCommand(deviceId: deviceId, command: command)
        }
    }
    
    func customBeepTime(deviceId: String, timeMs: Double, type: Double, frequencyHz: Double) throws -> Promise<Void> {
        return Promise.async {
            guard let manager = self.scannerManager else {
                throw NSError(domain: "ScannerBluetoothError", code: 2, userInfo: [NSLocalizedDescriptionKey: "ScannerBluetoothManager not initialized"])
            }
            
            let clampedTime = max(10, min(2540, Int(timeMs)))
            let clampedType = max(0, min(2, Int(type)))
            let clampedFreq = max(100, min(5200, Int(frequencyHz)))
            
            let temp1 = (clampedTime + 10) / 10
            let s1 = String(format: "%x", temp1)
            let s2 = String(clampedType)
            let temp3 = (clampedFreq - 100) / 20
            let s3 = String(format: "%x", temp3)
            
            let command = "$BUZZ#BK\(s1)\(s2)\(s3)"
            
            _ = try await manager.sendCommand(deviceId: deviceId, command: command)
        }
    }
    
    func powerOff(deviceId: String) throws -> Promise<Void> {
        return Promise.async {
            guard let manager = self.scannerManager else {
                throw NSError(domain: "ScannerBluetoothError", code: 2, userInfo: [NSLocalizedDescriptionKey: "ScannerBluetoothManager not initialized"])
            }
            
            _ = try await manager.sendCommand(deviceId: deviceId, command: "$POWER#OFF")
        }
    }
    
    // MARK: - Data Management
    
    func getStoredDataCount(deviceId: String) throws -> Promise<Double> {
        return Promise.async {
            guard let manager = self.scannerManager else {
                throw NSError(domain: "ScannerBluetoothError", code: 2, userInfo: [NSLocalizedDescriptionKey: "ScannerBluetoothManager not initialized"])
            }
            
            let response = try await manager.sendCommand(deviceId: deviceId, command: "%#+TCNT")
            // Parse the response to extract count
            // This would need proper parsing based on scanner response format
            return 0 // Placeholder
        }
    }
    
    func uploadStoredData(deviceId: String, clearAfterUpload: Bool) throws -> Promise<[ScanResult]> {
        return Promise.async {
            guard let manager = self.scannerManager else {
                throw NSError(domain: "ScannerBluetoothError", code: 2, userInfo: [NSLocalizedDescriptionKey: "ScannerBluetoothManager not initialized"])
            }
            
            let command = clearAfterUpload ? "%#TXMEM#C" : "%#TXMEM"
            _ = try await manager.sendCommand(deviceId: deviceId, command: command)
            
            // This would need proper implementation to collect multiple scan results
            return []
        }
    }
    
    func clearStoredData(deviceId: String) throws -> Promise<Void> {
        return Promise.async {
            guard let manager = self.scannerManager else {
                throw NSError(domain: "ScannerBluetoothError", code: 2, userInfo: [NSLocalizedDescriptionKey: "ScannerBluetoothManager not initialized"])
            }
            
            _ = try await manager.sendCommand(deviceId: deviceId, command: "%#*NEW*")
        }
    }
    
    // MARK: - Raw Command Interface
    
    func sendRawCommand(deviceId: String, command: String) throws -> Promise<String> {
        return Promise.async {
            guard let manager = self.scannerManager else {
                throw NSError(domain: "ScannerBluetoothError", code: 2, userInfo: [NSLocalizedDescriptionKey: "ScannerBluetoothManager not initialized"])
            }
            
            return try await manager.sendCommand(deviceId: deviceId, command: command)
        }
    }
}

File: android/src/main/java/com/margelo/nitro/bleprintandscan/HybridBleScanner.kt
package com.margelo.nitro.bleprintandscan

import android.app.Activity
import com.margelo.nitro.bleprintandscan.*
import com.margelo.nitro.core.Promise
import com.margelo.nitro.NitroModules
import com.bleprintandscan.ScannerBluetoothManager
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class HybridBleScanner: HybridBleScannerSpec() {
    
    private lateinit var scannerManager: ScannerBluetoothManager
    
    // MARK: - Scanner Connection Management
    
    override fun initializeScanner(): Promise<Unit> {
        return Promise.async {
            val appContext = NitroModules.applicationContext!!
            scannerManager = ScannerBluetoothManager(appContext)
            
            if (!scannerManager.isBluetoothSupported()) {
                throw Exception("Bluetooth is not supported on this device")
            }
            
            val activity = NitroModules.applicationContext!!.currentActivity as? Activity
                ?: throw Exception("Cannot initialize Scanner Bluetooth without an activity")
            
            val permissionsGranted = scannerManager.requestBluetoothPermissions(activity)
            if (!permissionsGranted) {
                throw Exception("Scanner Bluetooth permissions not granted")
            }
        }
    }
    
    override fun startScanningForScanners(onScannerFound: (Array<Device>) -> Unit): Promise<Unit> {
        return Promise.async {
            if (!scannerManager.isBluetoothEnabled()) {
                throw Exception("Please enable bluetooth")
            }
            
            scannerManager.startScanning { devices ->
                val deviceArray = devices.map { device ->
                    Device(device["id"]!!, device["name"]!!)
                }.toTypedArray()
                onScannerFound(deviceArray)
            }
        }
    }
    
    override fun suspendScanForScanners(): Promise<Unit> {
        return Promise.async {
            scannerManager.stopScanning()
        }
    }
    
    override fun connectToScanner(deviceId: String): Promise<Unit> {
        return Promise.async {
            if (!scannerManager.isBluetoothEnabled()) {
                throw Exception("Please enable bluetooth")
            }
            
            scannerManager.connectToScanner(deviceId)
        }
    }
    
    override fun disconnectFromScanner(deviceId: String): Promise<Unit> {
        return Promise.async {
            scannerManager.disconnect(deviceId)
        }
    }
    
    // MARK: - Multi-Scanner Management
    
    override fun isScannerConnected(deviceId: String): Promise<Boolean> {
        return Promise.async {
            scannerManager.isConnected(deviceId)
        }
    }
    
    override fun getConnectedScanners(): Promise<Array<Device>> {
        return Promise.async {
            val connectedScanners = scannerManager.getConnectedScanners()
            connectedScanners.map { scanner ->
                Device(scanner["id"]!!, scanner["name"]!!)
            }.toTypedArray()
        }
    }
    
    override fun disconnectAllScanners(): Promise<Unit> {
        return Promise.async {
            scannerManager.disconnectAllScanners()
        }
    }
    
    
    // MARK: - Scan Operations
    
    override fun startListening(deviceId: String, onScanResult: (ScanResult) -> Unit): Promise<Unit> {
        return Promise.async {
            if (!scannerManager.isConnected(deviceId)) {
                throw Exception("Scanner $deviceId is not connected")
            }
            
            scannerManager.startListening(deviceId) { scanData ->
                val scanResult = ScanResult(
                    data = scanData["data"]!!,
                    timestamp = scanData["timestamp"]!!,
                    deviceId = scanData["deviceId"]!!,
                    deviceName = scanData["deviceName"]!!
                )
                onScanResult(scanResult)
            }
        }
    }
    
    override fun stopListening(deviceId: String): Promise<Unit> {
        return Promise.async {
            scannerManager.stopListening(deviceId)
        }
    }
    
    override fun triggerScan(deviceId: String, duration: ScannerTrigger): Promise<Unit> {
        return Promise.async {
            if (!scannerManager.isConnected(deviceId)) {
                throw Exception("Scanner $deviceId is not connected")
            }
            
            val command = "%SCANTM#${duration.ordinal + 1}#"
            scannerManager.sendCommand(deviceId, command)
        }
    }
    
    // MARK: - Scanner Configuration
    
    override fun setScannerMode(deviceId: String, mode: ScannerMode): Promise<Unit> {
        return Promise.async {
            if (!scannerManager.isConnected(deviceId)) {
                throw Exception("Scanner $deviceId is not connected")
            }
            
            val command = when (mode) {
                ScannerMode.KEY_HOLD -> "%SCMD#00#"
                ScannerMode.CONTINUOUS -> "%SCMD#01#"
                ScannerMode.KEY_PULSE -> "%SCMD#02#"
                ScannerMode.HOST_TRIGGER -> "%SCMD#03#"
            }
            
            scannerManager.sendCommand(deviceId, command)
        }
    }
    
    override fun setBeepSettings(deviceId: String, settings: BeepSettings): Promise<Unit> {
        return Promise.async {
            if (!scannerManager.isConnected(deviceId)) {
                throw Exception("Scanner $deviceId is not connected")
            }
            
            val command = when (settings.volume) {
                BeepVolume.MUTE -> ScannerBluetoothManager.BEEP_MUTE
                BeepVolume.LOW -> ScannerBluetoothManager.BEEP_LOW_VOLUME
                BeepVolume.MIDDLE -> ScannerBluetoothManager.BEEP_MIDDLE_VOLUME
                BeepVolume.HIGH -> ScannerBluetoothManager.BEEP_HIGH_VOLUME
            }
            
            scannerManager.sendCommand(deviceId, command)
        }
    }
    
    override fun setPowerSettings(deviceId: String, settings: PowerSettings): Promise<Unit> {
        return Promise.async {
            if (!scannerManager.isConnected(deviceId)) {
                throw Exception("Scanner $deviceId is not connected")
            }
            
            val command = when (settings.sleepTimeMinutes.toInt()) {
                0 -> ScannerBluetoothManager.NEVER_SLEEP
                1 -> "\$RF#ST02"
                3 -> "\$RF#ST06"
                5 -> "\$RF#ST10"
                10 -> "\$RF#ST20"
                30 -> "\$RF#ST60"
                60 -> "\$RF#ST<0"
                120 -> "\$RF#STH0"
                else -> ScannerBluetoothManager.NEVER_SLEEP
            }
            
            scannerManager.sendCommand(deviceId, command)
        }
    }
    
    override fun setDataFormatSettings(deviceId: String, settings: DataFormatSettings): Promise<Unit> {
        return Promise.async {
            if (!scannerManager.isConnected(deviceId)) {
                throw Exception("Scanner $deviceId is not connected")
            }
            
            val commands = mutableListOf<String>()
            
            if (settings.prefixEnabled) {
                commands.add("\$DATA#2")
            }
            if (settings.suffixEnabled) {
                commands.add("\$DATA#1")
            }
            if (settings.hideBarcodePrefix) {
                commands.add("\$DATA#5")
            }
            if (settings.hideBarcodeContent) {
                commands.add("\$DATA#4")
            }
            if (settings.hideBarcodeSuffix) {
                commands.add("\$DATA#3")
            }
            
            for (command in commands) {
                scannerManager.sendCommand(deviceId, command)
            }
        }
    }
    
    override fun setTimestamp(deviceId: String, format: TimestampFormat, datetime: String?): Promise<Unit> {
        return Promise.async {
            if (!scannerManager.isConnected(deviceId)) {
                throw Exception("Scanner $deviceId is not connected")
            }
            
            val command = when (format) {
                TimestampFormat.DISABLED -> ScannerBluetoothManager.DISABLE_TIMESTAMP
                TimestampFormat.DATE_TIME, TimestampFormat.UNIX_TIMESTAMP -> ScannerBluetoothManager.ENABLE_TIMESTAMP
            }
            
            scannerManager.sendCommand(deviceId, command)
        }
    }
    
    // MARK: - Advanced Commands
    
    override fun restoreFactorySettings(deviceId: String): Promise<Unit> {
        return Promise.async {
            if (!scannerManager.isConnected(deviceId)) {
                throw Exception("Scanner $deviceId is not connected")
            }
            
            scannerManager.sendCommand(deviceId, ScannerBluetoothManager.RESTORE_FACTORY_SETTINGS)
        }
    }
    
    override fun customBeep(deviceId: String, level: Double): Promise<Unit> {
        return Promise.async {
            if (!scannerManager.isConnected(deviceId)) {
                throw Exception("Scanner $deviceId is not connected")
            }
            
            val clampedLevel = level.toInt().coerceIn(0, 26)
            val command = "\$BUZZ#B${(clampedLevel + 0x30).toChar()}"
            
            scannerManager.sendCommand(deviceId, command)
        }
    }
    
    override fun customBeepTime(deviceId: String, timeMs: Double, type: Double, frequencyHz: Double): Promise<Unit> {
        return Promise.async {
            if (!scannerManager.isConnected(deviceId)) {
                throw Exception("Scanner $deviceId is not connected")
            }
            
            val clampedTime = timeMs.toInt().coerceIn(10, 2540)
            val clampedType = type.toInt().coerceIn(0, 2)
            val clampedFreq = frequencyHz.toInt().coerceIn(100, 5200)
            
            val temp1 = (clampedTime + 10) / 10
            val s1 = temp1.toString(16)
            val s2 = clampedType.toString()
            val temp3 = (clampedFreq - 100) / 20
            val s3 = temp3.toString(16)
            
            val command = "\$BUZZ#BK$s1$s2$s3"
            
            scannerManager.sendCommand(deviceId, command)
        }
    }
    
    override fun powerOff(deviceId: String): Promise<Unit> {
        return Promise.async {
            if (!scannerManager.isConnected(deviceId)) {
                throw Exception("Scanner $deviceId is not connected")
            }
            
            scannerManager.sendCommand(deviceId, ScannerBluetoothManager.POWER_OFF)
        }
    }
    
    // MARK: - Data Management
    
    override fun getStoredDataCount(deviceId: String): Promise<Double> {
        return Promise.async {
            if (!scannerManager.isConnected(deviceId)) {
                throw Exception("Scanner $deviceId is not connected")
            }
            
            val response = scannerManager.sendCommand(deviceId, "%#+TCNT")
            // Parse the response to extract count
            // This would need proper parsing based on scanner response format
            0.0 // Placeholder
        }
    }
    
    override fun uploadStoredData(deviceId: String, clearAfterUpload: Boolean): Promise<Array<ScanResult>> {
        return Promise.async {
            if (!scannerManager.isConnected(deviceId)) {
                throw Exception("Scanner $deviceId is not connected")
            }
            
            val command = if (clearAfterUpload) "%#TXMEM#C" else "%#TXMEM"
            scannerManager.sendCommand(deviceId, command)
            
            // This would need proper implementation to collect multiple scan results
            arrayOf<ScanResult>()
        }
    }
    
    override fun clearStoredData(deviceId: String): Promise<Unit> {
        return Promise.async {
            if (!scannerManager.isConnected(deviceId)) {
                throw Exception("Scanner $deviceId is not connected")
            }
            
            scannerManager.sendCommand(deviceId, "%#*NEW*")
        }
    }
    
    // MARK: - Raw Command Interface
    
    override fun sendRawCommand(deviceId: String, command: String): Promise<String> {
        return Promise.async {
            if (!scannerManager.isConnected(deviceId)) {
                throw Exception("Scanner $deviceId is not connected")
            }
            
            scannerManager.sendCommand(deviceId, command)
        }
    }
}